File: solutions/0800-0899/875-koko-eating-bananas/875-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int left = 1;
        int right = 0;
        for(int pile : piles){
            right = Math.max(right, pile);
        }
        while(left <= right){
            int mid = left + (right-left)/2;
            long hour = eatBanana(piles, mid);
            if(hour <= (long) h){
                right = mid - 1;
            }else{ //hour > h Only increase left boundary when we eat slower than h
                left = mid + 1;
            }
        }
        return left;
    }
    public long eatBanana(int[] piles, int mid){
        long result = 0;
        for(int pile : piles){
            result += (long) (pile/mid + (pile % mid != 0 ? 1 : 0));
        }
        return result;
    }
}

File: solutions/0800-0899/875-koko-eating-bananas/875-koko-eating-bananas.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {     
        int left = 1, right = *max_element(piles.begin(), piles.end());

        while (left < right) {
            int middle = (left + right) / 2;
            int hourSpent = 0;

            for (int pile : piles) {
                hourSpent += pile / middle + (pile % middle != 0);
            }
            if (hourSpent <= h) {
                right = middle;
            } else {
                left = middle + 1;
            }
        }

        return right;
    }
};
